pkg/featurechangestopper/featurechangestopper.go

```diff
@@ -0,0 +1,140 @@
+func New(
+	startingTechPreviewState bool,
+	featureGateInformer configinformersv1.FeatureGateInformer,
+) *TechPreviewChangeStopper {
+	c := &TechPreviewChangeStopper{
+		startingTechPreviewState: startingTechPreviewState,
+		featureGateLister:        featureGateInformer.Lister(),
+		cacheSynced:              []cache.InformerSynced{featureGateInformer.Informer().HasSynced},
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "feature-gate-stopper"),
+	}
+
+	featureGateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.queue.Add("cluster")
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			c.queue.Add("cluster")
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.queue.Add("cluster")
+		},
+	})
+
+	return c
+}
+
+func (c *TechPreviewChangeStopper) syncHandler(ctx context.Context, shutdownFn context.CancelFunc) error {
+	featureGates, err := c.featureGateLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+	techPreviewNowSet := featureGates.Spec.FeatureSet == configv1.TechPreviewNoUpgrade
+	if techPreviewNowSet != c.startingTechPreviewState {
+		klog.Infof("TechPreviewNoUpgrade status changed from %v to %v, shutting down.", c.startingTechPreviewState, techPreviewNowSet)
+		if shutdownFn != nil {
+			shutdownFn()
+		}
+		return nil
+	}
+	return nil
+}
+
+func (c *TechPreviewChangeStopper) Run(runContext context.Context, shutdownFn context.CancelFunc) {
+	// don't let panics crash the process
+	defer utilruntime.HandleCrash()
+	// make sure the work queue is shutdown which will trigger workers to end
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting stop-on-techpreview-change controller")
+
+	// wait for your secondary caches to fill before starting your work
+	if !cache.WaitForCacheSync(runContext.Done(), c.cacheSynced...) {
+		return
+	}
+
+	// runWorker will loop until "something bad" happens.  The .Until will
+	// then rekick the worker after one second
+	go wait.UntilWithContext(runContext, func(ctx context.Context) {
+		c.runWorker(ctx, shutdownFn)
+	}, time.Second)
+
+	// wait until we're told to stop
+	<-runContext.Done()
+	klog.Infof("Shutting down stop-on-techpreview-change controller")
+}
+
+func (c *TechPreviewChangeStopper) runWorker(ctx context.Context, shutdownFn context.CancelFunc) {
+	// hot loop until we're told to stop.  processNextWorkItem will
+	// automatically wait until there's work available, so we don't worry
+	// about secondary waits
+	for c.processNextWorkItem(ctx, shutdownFn) {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false
+// when it's time to quit.
+func (c *TechPreviewChangeStopper) processNextWorkItem(ctx context.Context, shutdownFn context.CancelFunc) bool {
+	// pull the next work item from queue.  It should be a key we use to lookup
+	// something in a cache
+	key, quit := c.queue.Get()
+	if quit {
+		return false
```

Comment:
From [the `Get` docs][1]:

> If `shutdown = true`, the caller should end their goroutine.

By returning `false` here, you're asking `runWorker` to stop, but `UntilWithContext` will just launch a new round, and seems like that would get into an unblocked hotloop.  Should we use [`PollWithContext`][2] instead, and then plumb the `done` boolean back up to the poller?  Something like:

```golang
wait.UntilWithContext(runContext, 30 * time.Second, func(ctx context.Context) (bool, error) {
  // hot loop blah blah your `runWorker` comment...
  for {
    if done, err := c.processNextWorkItem(ctx, shutdownFn); done || err != nil {
      return done, err
    }
  }
})
```
[1]: https://pkg.go.dev/k8s.io/client-go/util/workqueue#Type.Get
[2]: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollWithContext

Reply:
the `Get` only returns `quit = true` after it has been shutdown. this happens *after* the context is closed, which is plumbed to `UntilWithContext` in `Run`, so it doesn't restart.

Reply:
> ...this happens _after_ the context is closed...

This is based on some non-local assumptions, and while I don't expect this particular controller to be all that churny, implicit, non-local assumptions are generally brittle.  I've floated kubernetes/community#6267 to tidy up some shutdown handling in the upstream template.

Reply:
Deferring this change; I'll file a follow-up CVO PR to pick up the stuff I'm suggesting in kubernetes/community#6267.